src/server_manager/web_app/cloud_accounts.ts

```diff
@@ -12,28 +12,129 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+import * as cloud from '../model/cloud';
 import * as digitalocean from '../model/digitalocean';
+import * as gcp from '../model/gcp';
+import {DigitalOceanAccount, ShadowboxSettings} from './digitalocean_account';
+import {GcpAccount} from './gcp_account';
 
-type DigitalOceanAccountFactory = (accessToken: string) => digitalocean.Account;
+export type AccountJson = {
+  digitalocean?: DigitalOceanAccountJson,
+  gcp?: GcpAccountJson,
+};
 
-export class CloudAccounts {
-  private readonly DIGITALOCEAN_TOKEN_STORAGE_KEY = 'LastDOToken';
+type DigitalOceanAccountJson = {
+  accessToken: string
+};
+
+type GcpAccountJson = {
+  refreshToken: string,
+};
+
+/**
+ * Manages connected cloud provider accounts.
+ */
+export class CloudAccounts implements cloud.CloudAccounts {
+  private readonly LEGACY_DIGITALOCEAN_STORAGE_KEY = 'LastDOToken';
+  private readonly ACCOUNTS_STORAGE_KEY = 'accounts-storage';
+
+  private digitalOceanAccount: DigitalOceanAccount = null;
+  private gcpAccount: GcpAccount = null;
 
   constructor(
-      private digitalOceanAccountFactory: DigitalOceanAccountFactory,
-      private storage = localStorage) { }
+      private shadowboxSettings: ShadowboxSettings,
+      private isDebugMode: boolean,
+      private storage = localStorage) {}
+
+  /**
+   * Loads the saved cloud accounts from disk.
+   *
+   * NOTE: This method must be called before calls to any other CloudAccounts
+   * methods. Failure to do so may result in accounts being disconnected or
+   * overwritten.
+   */
+  load(): void {
+    const accountJsonsString = this.storage.getItem(this.ACCOUNTS_STORAGE_KEY);
+
+    // Migrate any legacy DigitalOcean access token.
+    if (!accountJsonsString) {
+      const digitalOceanToken = this.storage.getItem(this.LEGACY_DIGITALOCEAN_STORAGE_KEY);
+      if (digitalOceanToken) {
+        this.digitalOceanAccount = this.createDigitalOceanAccount(digitalOceanToken);
+      }
+      this.save();
+    }
 
+    const accountJsons: AccountJson[] = accountJsonsString ? JSON.parse(accountJsonsString) : [];
```

Comment:
I thought more about the storage format and it seems we don't gain much from having all accounts in a shared list.

I'm thinking of a world where we can make the clouds more plug and play. Perhaps do things like accounts.digitalOcean().connect/get() with code for each cloud in their own unit (no need to make that in this PR). We would benefit from the storage for each cloud being independent.

So I suggest one more small change to the format:
- key: `accounts.digitalocean` with type DigitalOceanAccountJson[]
- key `accounts.gcp` with type GcpAccountJson[]

Then you can get rid of `AccountJson`.

Reply:
Done